tokio/src/sync/mpsc/bounded.rs

```diff
@@ -817,3 +975,120 @@ impl<T> fmt::Debug for Permit<'_, T> {
             .finish()
     }
 }
+
+// ===== impl Permit =====
+
+impl<T> OwnedPermit<T> {
+    /// Sends a value using the reserved capacity.
+    ///
+    /// Capacity for the message has already been reserved. The message is sent
+    /// to the receiver and the permit is consumed. The operation will succeed
+    /// even if the receiver half has been closed. See [`Receiver::close`] for
+    /// more details on performing a clean shutdown.
+    ///
+    /// Unlike [`Permit::send`], this method returns the [`Sender`] from which
+    /// the `OwnedPermit` was reserved.
+    ///
+    /// [`Receiver::close`]: Receiver::close
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use tokio::sync::mpsc;
+    ///
+    /// #[tokio::main]
+    /// async fn main() {
+    ///     let (tx, mut rx) = mpsc::channel(1);
+    ///
+    ///     // Reserve capacity
+    ///     let permit = tx.reserve_owned().await.unwrap();
+    ///
+    ///     // Send a message on the permit, returning the sender.
+    ///     let tx = permit.send(456);
+    ///
+    ///     // The value sent on the permit is received
+    ///     assert_eq!(rx.recv().await.unwrap(), 456);
+    ///
+    ///     // We may now reuse `tx` to send another message.
+    ///     tx.send(789).await.unwrap();
+    /// }
+    /// ```
+    pub fn send(mut self, value: T) -> Sender<T> {
+        let chan = self.chan.take().unwrap_or_else(|| {
+            unreachable!("OwnedPermit channel is only taken when the permit is moved")
+        });
```

Comment:
The unwrap here makes me a bit sad — not that I see any great ways to avoid it.

Reply:
I think it's necessary _if we want the `OwnedPermit` to be able to return the `Sender`_. Alternatively, we could remove `OwnedPermit::release` and change `OwnedPermit::send` to not return a `Sender`, and then we could remove the `Option` and just have `OwnedPermit` `mem::forget` itself to avoid running the `drop` code.  But, I think it's nice to be able to have one `Sender`, move it into an `OwnedPermit`, and then get the `Sender` back, without cloning or performing ref-count ops — in an async function, you could use this as a nice sort of type-state API...  

Also, removing the `Option` would complicate the implementation: since the `OwnedPermit` owns the `chan::Tx` rather than borrowing it, we would still want to decrement the reference count when it's consumed, but we wouldn't want to release capacity. `Permit` doesn't need to deal with this because it is only borrowing the `Tx`. So, we we would need to duplicate that code in `send` to ensure that the ref count is still decremented, or add a flag for whether or not the permit should be released.

:woman_shrugging:

Reply:
In principle we could avoid it in the same way as how `OwnedRwLockReadGuard` does it, but this is unsafe.

Reply:
Yeah, and I'm not sure if it's worth the unsafety here...what do you think?

Reply:
I have approved the PR. I'm fine either way.